source/writeread.c

```diff
@@ -0,0 +1,224 @@
+{
+    SSL *clientssl = NULL, *serverssl = NULL;
+    SSL_CTX *lsctx = NULL, *lcctx = NULL;
+    int ret = 1;
+    size_t i;
+    OSSL_TIME start, end;
+    char *sbuf, *cbuf;
+
+    /* Prepare client and server buffers. */
+    sbuf = OPENSSL_malloc(buf_size);
+    cbuf = OPENSSL_malloc(buf_size);
+    memset(cbuf, 0xaa, buf_size);
+
+    if (share_ctx == 1) {
+        lsctx = sctx;
+        lcctx = cctx;
+    } else {
+        if (!perflib_create_ssl_ctx_pair(smethod, cmethod, 0, 0,
+                                         &lsctx, &lcctx, cert, privkey)) {
+            fprintf(stderr, "Failed to create SSL_CTX pair\n");
+            err = 1;
+            return;
+        }
+    }
+
+    /* Setup connection. */
+    ret = perflib_create_ssl_objects(lsctx, lcctx, &serverssl, &clientssl,
+                                     NULL, NULL);
+    ret &= perflib_create_bare_ssl_connection(serverssl, clientssl,
+                                              SSL_ERROR_NONE);
+    if (!ret) {
+        err = 1;
+        return;
+    }
+
+    start = ossl_time_now();
+
+    for (i = 0; i < num_calls / threadcount; i++) {
+        size_t written = 0;
+        if (SSL_write_ex(clientssl, cbuf, buf_size, &written) <= 0) {
+            fprintf(stderr, "Failed to write data\n");
+            err = 1;
+            return;
+        }
+        size_t readbytes;
+        if (SSL_read_ex(serverssl, sbuf, buf_size, &readbytes) <= 0) {
+            fprintf(stderr, "Failed to read data\n");
+            err = 1;
+            return;
+        }
+        if (readbytes != written) {
+            fprintf(stderr, "Failed to read %ld bytes, got %ld\n", written, readbytes);
+            err = 1;
+            return;
+        }
+    }
+
+    end = ossl_time_now();
+    times[num] = ossl_time_subtract(end, start);
+
+    perflib_shutdown_ssl_connection(serverssl, clientssl);
+    if (share_ctx == 0) {
+        SSL_CTX_free(lsctx);
+        SSL_CTX_free(lcctx);
+    }
+    OPENSSL_free(sbuf);
+    OPENSSL_free(cbuf);
+}
+
+int main(int argc, char * const argv[])
+{
+    OSSL_TIME duration, ttime;
+    double avcalltime;
+    int ret = EXIT_FAILURE;
+    int i;
+    int terse = 0;
+    int opt;
+
+    /* Use TLS by default. */
+    smethod = TLS_server_method();
+    cmethod = TLS_client_method();
+
+    while ((opt = getopt(argc, argv, "tsdb:")) != -1) {
+        switch (opt) {
+        case 't':
+            terse = 1;
+            break;
+        case 's':
+            share_ctx = 0;
+            break;
+        case 'd':
+            smethod = DTLS_server_method();
+            cmethod = DTLS_client_method();
+            break;
+        case 'b':
+            buf_size = atoi(optarg);
+            if (buf_size < 1) {
+                printf("Buffer size argument must be > 0\n");
+                return EXIT_FAILURE;
+            }
+            break;
+        default:
+            printf("Usage: %s [-t] [-s] [-d] [-b size] certsdir threadcount\n",
```

Comment:
can you use `fprintf(stderr` here? just to keep it consistent with other perftools.

Reply:
I was following handshake.c but I now see there are differences. Using stderr in `main` as well is better, I'll update.